BrewR-EPFL/BrewR
# Develop/journey offlinemode

This pull request implements offline mode for the gallery screen and introduces significant improvements to the overall offline experience and real-time data updates. Below are the key changes:

Key Features Implemented:
Offline Coffee Shop Selection:

Users can now input a coffee shop name as plain text if they cannot find it in the dropdown list.
This is particularly useful for offline mode, where location suggestions might not be available.
Offline Journey Addition with Default Image:

When adding a journey offline, a default placeholder image is displayed.
Once the device is back online, the actual image is uploaded, and the gallery is automatically updated with the new image.
Real-Time Updates in Gallery:

Updated the getJourneys function in JourneysRepositoryFirestore to use Firestore's addSnapshotListener instead of get().
This ensures that any changes (whether offline or online) are reflected in the gallery in real-time.
Network Connectivity Check for Image Upload:

Introduced a new utility class, CheckConnection.kt, to check internet connectivity before attempting to upload images.
This prevents unnecessary upload attempts when offline and ensures smooth synchronization when back online.

Modified Files:
AddJourney.kt:

Added logic to show a default placeholder image in offline mode.
Incorporated a network connectivity check before image uploads.
Ensured proper synchronization of journeys when the device reconnects to the internet.
JourneyComponents.kt:

Updated UI components to gracefully handle and display placeholder images in offline mode.
JourneysRepositoryFirestore.kt:

Replaced get() with addSnapshotListener in the getJourneys method to enable real-time updates for the gallery.

CheckConnection.kt (New):
Added a utility class to check internet connectivity before performing network operations.

How to Test:
Test Offline Mode:

Add a journey while offline:
Verify that the journey appears in the gallery with a placeholder image.
Confirm that the gallery is updated with the actual image when the device reconnects to the internet.
Test Real-Time Updates:

Add or edit a journey while online and ensure the gallery updates in real-time without refreshing.
Test Coffee Shop Selection:

Add a journey and input the coffee shop name manually when dropdown suggestions are unavailable (e.g., in offline mode).
Test Connectivity Check:

Disable the internet and attempt to upload an image.
Confirm that the app uses the default image instead of trying to upload.

**Screenshot**
![at coffeeshop edit](https://github.com/user-attachments/assets/5706f820-b0cd-4ac6-9802-cbdc101c6c1e)
Users can input the name as a string in add screen in offline mode


![offlinemode photo](https://github.com/user-attachments/assets/e5d6b32b-72e3-4ebf-af60-cdfb3fc3f506)
Users see the update with default offline image when it is offline


![ChangeImplementedOnline](https://github.com/user-attachments/assets/846685cf-0e68-414e-981c-dfb7308eba2e)
Users can see real image that they uploaded when connected to online.
